### ISSUE TITLE
Compare CSLA to DevExpress XAF and other alternatives (if they exist)

### ISSUE THREAD

**Issue by aprokopyev** [MarimerLLC/cslaforum#844]
Dear Rockford,

Can you please help me to compare your CSLA to DevExpress XAF framework?

It seems these are the only business frameworks which survived till today.

I do not see any significant progress on any other frameworks mentioned at:
https://web.archive.org/web/20191016043646/http://forum.ideablade.com/forum_posts.asp?TID=767

They look obsolete and not compatible with the latest DotNet Core on Linux.

And I am not aware of any other framework.

Yours already supports the latest DotNet features and XAF is going too using at least Blazor as a GUI in about a year.

It seems like XAF is better in terms that it has a graphical designer for the whole business model while CSLA needs a lot of coding and declarative definition of everything except Microsoft GUI designers if used at all.

Is XAF much more productive?

Are there any other business framework alternatives for DotNet Core ?

Can you please highlight the strongest features of CSLA over XAF except price?

Best Regards,
Alex

**Comment by BlagoCuljak** [comment]
XAF is an ORM, like eg. Entity Framework. CSLA is much more.

**Comment by aprokopyev** [comment]
Most likely you confuse XAF (which is a Business Application Framework) and XPO (which is an ORM).

https://www.devexpress.com/products/net/application_framework/

As for me I prefer nHibernate as ORM and will continue to use nHibernate even if XAF will be choosen as a framework.

**Comment by BlagoCuljak** [comment]
@aprokopyev yes, you are right, my bad!

**Comment by rockfordlhotka** [comment]
I've never looked at or used XAF, so I can't compare.

It is true that, to my knowledge, no other framework focused on business logic has been around for the 22+ years that CSLA has existed. Some folks, like IdeaBlade, created competitors that were around for a few years here and there. But it is difficult to make money off this sort of thing, at least not the kind of money necessary to run a company, so they've all come and gone over time.

**Comment by Chicagoan2016** [comment]
My two cents, 
Csla is a business layer framework, Rocky does provide some controls for easy integration with different UI 'technologies' like WPF, Winforms, MVC etc.
I have never worked with XAF only Csla. 
How good is XAF's business rules engine? that's one of the strongest aspects of Csla.
Another thing I would be looking at, what kinda object stereotypes does XAF support?

Regards

**Comment by Chicagoan2016** [comment]
@rockfordlhotka , one of your good friends from CA (currently Nashville) used to sell an excellent framework :), I was a big fan but unfortunately he sold his company and I am not sure if he still sells his framework.

Kind Regards

**Comment by rockfordlhotka** [comment]
@Chicagoan2016 yes, that's true 😄 And another CA-based friend had NetTiers (??) or something like that.

There've been a few off and on over the years to be sure, each with their own philosophy and set of patterns.

**Comment by Chicagoan2016** [comment]
@rockfordlhotka ,I always thought your Nashville friend should have integrated his framework with Csla.
In my opinion that would have been a great combination.

Regards

**Comment by aprokopyev** [comment]
> How good is XAF's business rules engine? that's one of the strongest aspects of Csla.

I am not sure if XAF validation covers what you mention:
https://docs.devexpress.com/eXpressAppFramework/113684/concepts/extra-modules/validation/validation-module-overview

I did not use CSLA yet except a few simple tries.
If I understood correctly CSLA remembers a chain of user actions towards to business logic, those can be  undone if needed. Each action is validated for business rules conformance.

**Comment by aprokopyev** [comment]
!""(https://hsto.org/storage2/da8/6e0/342/da86e034271efe840beac50cd0054f52.png)

**Comment by aprokopyev** [comment]
> I've never looked at or used XAF, so I can't compare.
> But it is difficult to make money off this sort of thing, at least not the kind of money necessary to run a company, so they've all come and gone over time.

Perpetual (with only 1 year of upgrades) DevExpress XAF license costs about $2K initially and then $1K/year for upgrades if needed per each developer. End users applications based on XAF are free from DevExpress license charges.

**Comment by aprokopyev** [comment]
I wonder if there is any sense to try to use CSLA as object space provider for DevExpress XAF:

https://www.devexpress.com/Support/Center/Question/Details/E411/how-to-customize-the-underlying-database-provider-options-and-data-access-behavior-in-xaf

XAF Object Space -> CSLA -> nHibernate

**Comment by angtianqiang** [comment]
@aprokopyev I've been using CSLA for the business layer, DevExpress for the interface (WPF), and a framework that works great

**Comment by aprokopyev** [comment]
Most likely you had to manually design WPF forms or use a specific custom or third party generator for them. XAF does rendering automatically from a model definition. Currently to WinForms and WebForms, in about a year it is going to support also Blazor rendering.

Just switch Win/Web/Blazor, XAF will render it at runtime by a choosen GUI API.
No any manual changes at least if you customized GUI components, it is like a RAD magic.

**Comment by Chicagoan2016** [comment]
> Most likely you had to manually design WPF forms or used a custom or third party generator for them. On the other hand XAF does rendering to GUI forms automatically from a model definition without a need to design forms manually, the forms are still created dynamically by XAF at runtime.
> 
> Currently WinForms and WebForms on Windows are supported, but in about a year it is going to support also Blazor server side rendering (from model to web pages) even on a Linux server.
> 
> Just switch an option like Win/Web/Blazor and XAF will render your data at runtime by a choosen GUI API ruled by your model definition. No any manual changes needed at least unless you customized your own GUI components, it is like a RAD magic.

It starts to sound like an advertisement for XAF framework

**Comment by aprokopyev** [comment]
Actually I am not affiliated with DevExpress, they even do not have even an affiliate program for reselling and I did not place any links to any shop.

Though I have used to purchase a DevExpress subscription earlier and I am very happy with their GUI components, never seen anything more advanced and convenient at the same time.

**Comment by aprokopyev** [comment]
Please let me know how much CSLA is portable?

1) Can CSLA work on Mono under OpenBSD?
2) Can CSLA interact remotely crossplatform like:
CSLA tierX on DotNet v4.8 under Windows ->  CSLA tierY on DotNet Core v3x under Linux -> CSLA tierZ on Mono v6x under OpenBSD  
?

**Comment by rockfordlhotka** [comment]
I haven't tested all those platforms/scenarios, but in general the answer should be yes.

I _know_ it works on

1. mono (Xamarin sits on mono) on iOS/Android/Mac
1. .NET Core on Windows and Linux (amd64, arm32, arm64)
1. mono on Linux (Ubuntu, Debian, and others)
1. mono on Mac (which is a variant of OpenBSD right?)

Basically, if the platform can host .NET Standard 2.0 code then it will run CSLA - `Csla.dll` is a netstandard2.0 assembly.

**Comment by rockfordlhotka** [comment]
In terms of spanning platforms from client to server, that works across any platform combination.

**Comment by Chicagoan2016** [comment]
> In terms of spanning platforms from client to server via the data portal, that works across any platform combination.
> 
> Nearly everything I've been doing for the past 12-18 months involves running server-side code in Linux containers in Docker/Kubernetes/etc. and building client apps on Windows, Android, and now WebAssembly.

@rockfordlhotka , I am really interested in getting to know about this. I wonder if  you get some down time and may be write a blog about this :).
Kind Regards

**Comment by rockfordlhotka** [comment]
My goal, @Chicagoan2016, is to finish the Blazor book and then update the data portal book to include hosting in containers and using the "new" CSLA features to do routing to different containers.

**Comment by Chicagoan2016** [comment]
> My goal, @Chicagoan2016, is to finish the Blazor book and then update the data portal book to include hosting in containers and using the "new" CSLA features to do routing to different containers.

I look forward to reading the Blazor book, how many books are available on csla store regarding Csla V 5.X ?  I know the introduction is free to download.

Kind Regards

**Comment by rockfordlhotka** [comment]
Right now only the free introduction book is online. Blazor will be book number 2.
